Document how to develop DevTools tooling (#159)
servo/book: src/hacking/developing-devtools.md

The terminal window now contains full debug server logs; copy them to somewhere for further analysis.

## Capturing and processing protocol traffic

We have seen a simple way of obtaining message logs from Servo and Firefox.
However, this soon turns complex when wanting to compare logs between the two due to the different formats or performing queries on them.
There is a small script to make this process easier: [`etc/devtools_parser.py`](https://github.com/servo/servo/blob/main/etc/devtools_parser.py).

It is based on [Wireshark](https://www.wireshark.org/), a powerful network packet analyzer; more specifically it's cli, `tshark`.
It is configured to log packets send in your local network on the port that the devtools server is running, so it can read the payloads from them which contain the JSON packets from the DevTools protocol.

**`tshark` needs to be installed** for the script to work.
On [Setting up your environment](https://book.servo.org/hacking/setting-up-your-environment.html) the dependencies for many distros already include this, but check [the official install guide](https://tshark.dev/setup/install) if you are missing the package.

> Additionally, make sure to [set up a Firefox profile for debugging](https://book.servo.org/hacking/developing-devtools.html#on-the-first-run).

Reviewer: ```suggestion
Additionally, make sure to [set up a Firefox profile for debugging](#on-the-first-run).
```

### Capture a session

1. Run either Servo or Firefox with the DevTools server enabled:

```sh
./mach run --devtools 6080
firefox --new-instance --start-debugger-server 6080 -P devtools-testing
```

2. In another terminal, start the script in capture mode (`-w`), specifying the same port as before:

```
./etc/devtools_parser.py -p 6080 -w capture.pcap
```

3. Connect from `about:debugging` following [the same steps](https://book.servo.org/hacking/using-devtools.html#connecting-to-servo).
4. Perform any actions you want to record.
5. Press `Ctrl-C` in the terminal running the parser to stop the recording. This will do two things:

>     - Save a `.pcap` file where specified by the `-w` flag. This is a binary file format for Wireshark, but we can read it later with the same tool.

Reviewer: ```suggestion
    - Save the results to a `.pcap` file specified by the `-w` flag. 
      This is a binary file format for Wireshark, but we can read it later with the same tool.
```

    - Print the message log. There are two modes: the regular one, where it prints the messages in a friendly way, and `--json`, which emits newline-separated JSON with each message.
6. You can now close Servo or Firefox.

### Read a capture

It is useful to save multiple captures and compare them later.

> While `tshark` saves them by default in the `.pcap` format, we can use the same script to get a better output from them.

Reviewer: ```suggestion
While `tshark` saves them by default in the `.pcap` format, we can use the same script to get better output from them.
```
